loki/transformations/utilities.py

```diff
@@ -32,10 +35,167 @@
+    'modify_variable_declarations', 'RemoveDuplicateArgs',
 ]
 
 
+class RemoveDuplicateArgs(Transformation):
+    """
+    Transformation to remove duplicate arguments for both caller
+    and callee. 
+
+    .. warning::
+        this won't work properly for multiple calls to the same routine
+        with differing duplicate arguments
+
+    Parameters
+    ----------
+    recurse_to_kernels : bool, optional
+        Remove duplicate arguments only at the driver level or recurse to
+        (nested) kernels (Default: `True`).
+    rename_common : bool, optional
+        Try to rename duplicate arguments by finding a common pattern
+        in those names (Default: `False`).
+    """
+
+    # This trafo only operates on procedures
+    item_filter = (ProcedureItem,)
+
+    def __init__(self, recurse_to_kernels=True, rename_common=False):
+        self.recurse_to_kernels = recurse_to_kernels
+        self.rename_common = rename_common
+
+    def transform_subroutine(self, routine, **kwargs):
+        role = kwargs['role']
+        if role == 'driver' or self.recurse_to_kernels:
+            remove_duplicate_args(routine, rename_common=self.rename_common)
+
+def remove_duplicate_args(routine, rename_common=False):
+    """
+    Utility to remove duplicate arguments for both caller
+    and callee. 
+
+    .. warning::
+        this won't work properly for multiple calls to the same routine
+        with differing duplicate arguments
+
+    Parameters
+    ----------
+    routine : :any:`Subroutine`
+        The subroutine to be transformed.
+    rename_common : bool, optional
+        Try to rename duplicate arguments by finding a common pattern
+        in those names (Default: `False`).
+    """
+
+    def remove_duplicate_args_call(call):
+        arg_map = {}
+        for routine_arg, call_arg in call.arg_iter():
+            arg_map.setdefault(call_arg, []).append(routine_arg)
+        # filter duplicate kwargs (comparing to the other kwarguments)
+        _new_kwargs = as_tuple(list(_)[0] for g, _ in it.groupby(call.kwarguments, key=lambda x: x[1]))
+        # filter duplicate kwargs (comparing to the arguments)
+        new_kwargs = tuple(kwarg for kwarg in _new_kwargs if kwarg[1] not in call.arguments)
+        # (filter duplicate arguments and) update call
+        call._update(arguments=as_tuple(dict.fromkeys(call.arguments)), kwarguments=new_kwargs)
+        return arg_map
+
+    def modify_callee(callee, callee_arg_map):
+        combine = [routine_args for call_arg, routine_args in callee_arg_map.items() if len(routine_args) > 1]
+        matches = [os.path.commonprefix([str(elem.name) for elem in args]).rstrip('_')
+                or os.path.commonprefix([str(elem.name)[::-1] for elem in args]).rstrip('_')[::-1] for args in combine]
+        rename_common_map = {combine[i][0].name: matches[i] for i in range(len(combine)) if matches[i] != ''}\
+                if rename_common else {}
```

Comment:
For readability, I would prefer not to do use an inline `if` here. Wrapping this in proper conditional branches also avoids building `matches` if it's not required.

Secondly, the common pattern for `combine[i]: matches[i] for i in range(len(combine))` is to use `zip`:
```suggestion
        if rename_common:
            matches = [
                os.path.commonprefix([str(elem.name) for elem in args]).rstrip('_') or
                os.path.commonprefix([str(elem.name)[::-1] for elem in args]).rstrip('_')[::-1]
                for args in combine
            ]
            rename_common_map = {c[0].name: m for c, m in zip(combine, matches) if m}
        else:
            rename_common_map = {}
```
```